src/order/heyting/hom.lean

```diff
@@ -0,0 +1,405 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import order.hom.complete_lattice
+
+/-!
+# Heyting algebra morphisms
+
+A Heyting homomorphism between two Heyting algebras is a bounded lattice homomorphism that preserves
+Heyting implication.
+
+We use the `fun_like` design, so each type of morphisms has a companion typeclass which is meant to
+be satisfied by itself and all stricter types.
+
+## Types of morphisms
+
+* `heyting_hom`: Heyting homomorphisms.
+* `coheyting_hom`: Co-Heyting homomorphisms.
+* `biheyting_hom`: Bi-Heyting homomorphisms.
+
+## Typeclasses
+
+* `heyting_hom_class`
+* `coheyting_hom_class`
+* `biheyting_hom_class`
+-/
+
+open function
+
+variables {F α β γ δ : Type*}
+
+/-- The type of Heyting homomorphisms from `α` to `β`. Bounded lattice homomorphisms that preserve
+Heyting implication. -/
+@[protect_proj]
+structure heyting_hom (α β : Type*) [heyting_algebra α] [heyting_algebra β]
+  extends lattice_hom α β :=
+(map_bot' : to_fun ⊥ = ⊥)
```

Comment:
Do you have a source (or proof) that bot needs to be preserved? I don't think I've seen this condition in a definition of Heyting algebra homs before

Reply:
Yes, Borceux above does assume it. It's similar to how you can construct `⊤` from `generalized_heyting_algebra`. I think `f = ⊤` works as a counterexample?

Reply:
If I remember right, all (nontrivial?) linear orders are Heyting algebras (with `a → b` given by `⊤` if `a ≤ b` and `a` otherwise), and for a counterexample you can take the map taking the linear order with two elements and adding a new bottom element. So `map_bot` definitely is needed.

Reply:
We have [this in mathlib](https://leanprover-community.github.io/mathlib_docs/order/heyting/basic.html#linear_order.to_biheyting_algebra)!